### ISSUE TITLE
Floating point arithmetic error when converting units of multiple powers

### ISSUE THREAD

**Issue by rocco8773** [astropy/astropy#9721]
Over in PlasmaPy we're running into an issue where converting between SI prefixes for the same unit does not always produce the expected result (see [here](https://github.com/PlasmaPy/PlasmaPy/pull/722#issuecomment-558917491)).  The order-zero conversion is always correct (i.e. `(100 * u.cm).to(u.m) == 1. * u.m` is `True`), but when we have powers of units (i.e. `u.cm ** 2`, `u.cm ** 5`, etc.) then the conversion is not always exact.

Here are two screenshots the illustrates the issue more clearly...

1. here I'm always converting to `1.0 * u.m ** power` from `u.cm ** power`

    <img width="748" alt="image" src="https://user-images.githubusercontent.com/29869348/70101480-e0b8eb80-15e9-11ea-8ff6-89cf3bd249de.png">

1. here I'm always converting to `1.e21 * u.m ** power` from `u.cm ** power`

    <img width="755" alt="image" src="https://user-images.githubusercontent.com/29869348/70101532-0cd46c80-15ea-11ea-975a-48d0d90b5dcf.png">

----
* This happens for any conversion between SI prefixes.
* When the difference occurs, it always seems to be on the order of `10 ** 16`, irregardless of which units we're converting from and to.

**Comment by mhvk** [comment]
I think these are simply floating point rounding issues: in a binary system, one cannot exactly store any negative power of 10 (like `0.1`), and at least in your top example the differences are close to the cut-off for floating point arithmetic:
```
np.finfo(float).eps
# 2.220446049250313e-16
# = 2**-52
```
Note that these precisions are fractional ones; to check whether something is close to best precision, you can use `np.testing.assert_array_max_ulp`; `np.spacing` gives generically what the closest till representable number is.

**Comment by mhvk** [comment]
p.s. Thanks for the very detailed report!!

**Comment by rocco8773** [comment]
@mhvk Thanks for the reply.  I gathered this would be the response.  While the error is relatively small, it's having an affect on a calculation we're trying to do.

I've noticed that `Quantities` have `dtype=np.float64`, is it possible to recast them as `np.float128`?  Since `np.finfo(np.float128).eps == 1.084202172485504434e-19`, maybe this error can be mitigate a bit more.

**Comment by mhvk** [comment]
@rocco8773 - it is not just that it is small, but that it is essentially unavoidable! What particular application do you have? Often one can avoid to need all 16 digits of precision that regular doubles give one.

In any case, to more directly answer, yes, you can use `float128` if you wish (just pass the relevant `dtype`), though it gives only 3 digits more precision, as it is really just `float80` stored in 128 bits (on intel/amd at least).

**Comment by rocco8773** [comment]
@mhvk I have a function that's calculating a plasma parameter.  We've constructed a decorator that basically does unit conversions into and out of the function, so the user can pass-in any equivalent units and we convert the parameter to the units the function expects.  Unfortunately, [this function appears to be very sensitive to one of the input parameters (density `u.m ** -3`)](https://github.com/PlasmaPy/PlasmaPy/pull/722#issuecomment-558917491).  Fortunately, this input parameter gets normalized inside the function, so I can probably hold-off on the conversion until after the normalization, then the conversion is only with units to the power of 1.

**Comment by mhvk** [comment]
Looking at the code, it is very difficult to see how the unit conversion could be the underlying problem. Is perhaps the function converging badly? Or that the `Fermi_integral` is not precise enough for small `alpha`? It may be an idea to calculate the actual expected value carefully (there must be some near-ideal expansion to compare against!). Or add a debug print statement in `residual` to see what values are actually tried, and what the integral looks like.

**Comment by rocco8773** [comment]
I agree! I didn't write the function, so my familiarity with it is poor at best.  There's obvious a high level of sensitivity in the fitting algorithm.  I discovered this conversion error in my initial investigation, so I wanted to get some info on it as I delved into the other possibilities.

For now I think my questions are satisfied on this posted issue.  Feel free to close the issue.  If I discover more penitent info in my investigation, then I'll re-open it.

Thanks for all your help!
